Added Unpacker and Packer for Emon Engine.

HOKORISAMA opened the pull request:
Hey there! So I was working on Emon Engine from a long time and I've successfully made packer and unpacker for all file formats as well Image convertor, but the issue is we need a little metadata to create file that is a 40 byte key that is stored in the file itself and differs for each game engine file. 
To pack script files that is nz_scene.eme we need to change the SubType to 3 and 4 for image file(a custom bmp image).
LzssFrameSize is constant that is 4096, LzssInitPos is 4078(calculated) but initially is (18).

We can ask the user if he/she wants to create a script file archive or a Image archive and set SubType accordingly but if we do not want to use any external metadata file we could use the original file to read the key and use that for file creation.


Let me know how you feel about this so we can approach that way.
(btw I am done with image convertor as well, do have any plans to make GalArc support image conversion as well?)


(NOTE: This is a rough implementation and needs improvement to make it flexible for different scenarios.)

HOKORISAMA commented:
I am done with GSPack as well and will create a pull request soon.

detached64 commented:
> We can ask the user if he/she wants to create a script file archive or a Image archive and set SubType accordingly

Maybe we can check the input folder name or file extensions? At least we can use the radio button.....:-)

> if we do not want to use any external metadata file we could use the original file to read the key and use that for file creation.

I've also researched emon emgine for a while. As far as I know, the first 8 bytes of the 40 bytes key defines encryption method, and other 32 bytes is what we use in decryption. Those defining method maybe have these valid values according to GARbro: 1, 2, 4, 8. So if we set them to 0, the game perhaps does no decryption. In that case we don't need encryption algorithm as well as the 40 bytes key.
[ref](https://github.com/morkt/GARbro/issues/614#issuecomment-1904912254)

> btw I am done with image convertor as well, do have any plans to make GalArc support image conversion as well?

Well, I'm not prepared to introduce image conversion function into GalArc temporarily. It's a complicated work I must say, not as trivial as script de/encryption. Besides, GalArc focuses on Archive, where other function should play a subsidiary role, and I don't want "the sauce is better than the fish".

Maybe I'll write another application for image conversion with some new features like preview, batch conversion,.etc.

detached64 commented:
Merged. Thanks for your contribution.

HOKORISAMA commented:
Got it, btw which game you used to reaserch for emon engine?

detached64 commented:
I used [瞳の烙淫～淫縛の牝奴隷～](https://vndb.org/v7363).

detached64 commented:
I've implemented unpacking and packing method of Emon Engine. I use file extension to judge type. Tested on 瞳の烙淫～淫縛の牝奴隷～.

HOKORISAMA commented:
> I've implemented unpacking and packing method of Emon Engine. I use file extension to judge type. Tested on 瞳の烙淫～淫縛の牝奴隷～. Besides, set key to 0 works well, so I deleted encryption methods. btw which game you used to research for gspack?

You can use any [Aurobahn](https://vndb.org/p1617) game, all are using same engine, but I used [Kanjuku Rouge Drip](https://vndb.org/v10815) and [Yarimoku Nanapa Beach](https://vndb.org/v14366).

I'm also trying to make a script extarctor to translate the games using this engine but I am nowhere near a functional tool.
The script files are a mess.

detached64 commented:
Well, maybe need opcode or something to parse script files, not familiar with that.
